wifi.go

```diff
@@ -0,0 +1,134 @@
+// to a wireless network.
+type WiFiInfo struct {
+	SSID         string `json:"ssid,omitempty"`
+	Passphrase   string `json:"passphrase,omitempty"`
+	SecurityType int    `json:"securityType,omitempty"`
+}
+
+const (
+	SecurityTypeNone int = 0
+	SecurityTypeWEP  int = 1
+	SecurityTypePSK  int = 2
+)
+
+// SetWiFiInfo encrypts and updates a Key Light's WiFi configuration.
+func (c *Client) SetWiFiInfo(ctx context.Context, wifi *WiFiInfo, device *Device) error {
+	b, err := json.Marshal(wifi)
+	if err != nil {
+		return err
+	}
+
+	blen := len(b)
+	padlen := aes.BlockSize - (blen % aes.BlockSize)
+	pad := bytes.Repeat([]byte{0}, padlen)
+
+	plaintext := append(b, pad...)
+	key := aesKey(device.HardwareBoardType, device.FirmwareBuildNumber)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return err
+	}
+
+	ciphertext := make([]byte, aes.BlockSize+len(plaintext))
+	iv := ciphertext[:aes.BlockSize]
+	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
+		return err
+	}
+
+	mode := cipher.NewCBCEncrypter(block, iv)
+	mode.CryptBlocks(ciphertext[aes.BlockSize:], plaintext)
+
+	return c.do(ctx, http.MethodPost, "/elgato/wifi-info", bytes.NewReader(ciphertext), nil)
+}
+
+// GetWiFiInfo fetches and decrypts a Key Light's WiFi configuration.
+func (c *Client) GetWiFiInfo(ctx context.Context, device *Device) (*WiFiInfo, error) {
+	ciphertext, err := c.doRaw(ctx, http.MethodGet, "/elgato/wifi-info", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	key := aesKey(device.HardwareBoardType, device.FirmwareBuildNumber)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ciphertext) < aes.BlockSize {
+		return nil, fmt.Errorf("ciphertext too short")
+	}
+	iv := ciphertext[:aes.BlockSize]
+	ciphertext = ciphertext[aes.BlockSize:]
+
+	if len(ciphertext)%aes.BlockSize != 0 {
+		return nil, fmt.Errorf("ciphertext is not a multiple of the block size")
+	}
+
+	mode := cipher.NewCBCDecrypter(block, iv)
+	mode.CryptBlocks(ciphertext, ciphertext)
+
+	idx := bytes.IndexByte(ciphertext, 0)
+	if idx < 0 {
+		idx = len(ciphertext)
+	}
+
+	wifi := &WiFiInfo{}
+	err = json.Unmarshal(ciphertext[:idx], wifi)
+	if err != nil {
+		return nil, err
+	}
+
+	return wifi, nil
+}
+
+// doRaw performs an HTTP request with the input parameters, returning
+// the raw body is returned.
+func (c *Client) doRaw(ctx context.Context, method, path string, body io.Reader) ([]byte, error) {
+	u := *c.u
+	u.Path = path
+
+	req, err := http.NewRequestWithContext(ctx, method, u.String(), body)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := c.c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("keylight: device returned HTTP %d", res.StatusCode)
+	}
+
+	return ioutil.ReadAll(res.Body)
+}
+
+// aesKey returns the AES key based on the device's board type and firmware build number.
+func aesKey(boardType, firmwareBuildNumber int) []byte {
+	return []byte{
```

Comment:
Any links you can share for how this is derived? Or did you just get this from the web UI JS?